File: Job Sequencing Problem - GFG/job-sequencing-problem.cpp
// { Driver Code Starts
// Program to find the maximum profit job sequence from a given array 
// of jobs with deadlines and profits 
#include<bits/stdc++.h>
using namespace std; 

// A structure to represent a job 
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
}; 


 // } Driver Code Ends
/*
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
};
*/

class Solution 
{
    struct comp{
        public:
        bool operator()(Job j1, Job j2){
            return j1.profit < j2.profit; // whichever has a higher profit should be on top
        }
    };
    public:
    //Function to find the maximum profit and the number of jobs done.
    vector<int> JobScheduling(Job arr[], int n) 
    { 
        // your code here
        // DRY RUN
        
        priority_queue <Job,vector<Job>,comp> pq,copy_pq;
        
        int maxDead=-1;
        for(int i=0;i<n;i++){
            pq.push(arr[i]); // {(3,1,40), (4,1,30) , (1,4,20), (2,1,10)}
        
            maxDead = max(maxDead,arr[i].dead);
        }
        
        vector<int> slots(maxDead,-1);
        // copy_pq = pq;
        // while(!copy_pq.empty()){
        //     cout<< "profit: "<<copy_pq.top().profit<<" deadline:"<<copy_pq.top().dead<<endl;
        //     copy_pq.pop();
        // }
        int doneJobs=0,profit =0;
        while(!pq.empty()){
            auto currentJob = pq.top(); // {3,1,40} -> {4,1,30} -> {1,4,20} -> {2,1,10}
            pq.pop();
            for(int i=currentJob.dead-1;i>=0;--i){
                if(slots[i]==-1){
                    slots[i] = currentJob.id;
                    profit+= currentJob.profit;
                    doneJobs++;
                    break;
                }
            }
        }
        
        
        return {doneJobs,profit};
    } 
    
};

// { Driver Code Starts.
// Driver program to test methods 
int main() 
{ 
    int t;
    //testcases
    cin >> t;
    
    while(t--){
        int n;
        
        //size of array
        cin >> n;
        Job arr[n];
        
        //adding id, deadline, profit
        for(int i = 0;i<n;i++){
                int x, y, z;
                cin >> x >> y >> z;
                arr[i].id = x;
                arr[i].dead = y;
                arr[i].profit = z;
        }
        Solution ob;
        //function call
        vector<int> ans = ob.JobScheduling(arr, n);
        cout<<ans[0]<<" "<<ans[1]<<endl;
    }
	return 0; 
}


  // } Driver Code Ends

File: Job Sequencing Problem - GFG/job-sequencing-problem.py
#User function Template for python3
import heapq
class Solution:
    
    #Function to find the maximum profit and the number of jobs done.
    def JobScheduling(self,Jobs,n):
        arr = [(job.id,job.deadline,job.profit) for job in Jobs]
        arr.sort(key=lambda x: (x[1]))
        totProfit= 0
        # initialise the result array and maxHeap
        result = []
        maxHeap = []
        # starting the iteration from the end
        for i in range(n - 1, -1, -1):
    
            # calculate slots between two deadlines
            if i == 0:
                slots_available = arr[i][1]
            else:
                slots_available = arr[i][1] - arr[i - 1][1]
    
            # include the profit of job(as priority), deadline
            # and job_id in maxHeap
            # note we push negative value in maxHeap to convert
            # min heap to max heap in python
            heapq.heappush(maxHeap, (-arr[i][2], arr[i][1], arr[i][0]))
    
            while slots_available and maxHeap:
    
                # get the job with max_profit
                profit, deadline, job_id = heapq.heappop(maxHeap)
    
                # reduce the slots
                slots_available -= 1
    
                # include the job in the result array
                result.append([job_id, deadline,profit])
                totProfit+=-profit
    
        # jobs included might be shuffled
        # sort the result array by their deadlines
        # result.sort(key=lambda x: x[1])
        return [len(result),totProfit]
        
        # code here

#{ 
#  Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

#Contributed by : Nagendra Jha
class Job:
    '''
    Job class which stores profit and deadline.
    '''
    def __init__(self,profit=0,deadline=0):
        self.profit = profit
        self.deadline = deadline
        self.id = 0

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        n = int(input())
        info = list(map(int,input().strip().split()))
        Jobs = [Job() for i in range(n)]
        for i in range(n):
            Jobs[i].id = info[3*i]
            Jobs[i].deadline = info[3 * i + 1]
            Jobs[i].profit=info[3*i+2]
        ob = Solution()
        res = ob.JobScheduling(Jobs,n)
        print (res[0], end=" ")
        print (res[1])
# } Driver Code Ends